py/DREAM/DREAMTask.py

```diff
@@ -0,0 +1,92 @@
+import os
+import pathlib
+import subprocess
+import tempfile
+import time
+
+from . DREAMException import DREAMException
+from . DREAMOutput import DREAMOutput
+from . DREAMSettings import DREAMSettings
+from . DREAMErrorOutput import DREAMErrorOutput
+from subprocess import TimeoutExpired
+
+class DREAMTask:
+    def __init__(self, settings, outfile=None, quiet=False, timeout=None, DREAMPATH=None):
+        self.deleteOutput = False
+        if outfile is None:
+            self.deleteOutput = True
+            self.outfile = next(tempfile._get_candidate_names())+'.h5'
+        else:
+            self.outfile = outfile
+
+        self.infile = None
+        if isinstance(settings, DREAMSettings):
+            self.infile = next(tempfile._get_candidate_names())+'.h5'
+            settings.output.setFilename(self.outfile)
+            settings.save(self.infile)
+        else:
+            self.infile = settings
+
+        self.settings = settings
+        self.errorOnExit = 0
+        self.p = None
+        self.obj = None
+        self.stderr_data = None
+        self.DREAMPATH = DREAMPATH
+        self.quiet = quiet
+        self.timeout = timeout if timeout is not None else float('inf')
+
+    def run(self):
+        self.startTime= time.time()
+        if self.p != None: #if process is already created then we can safely od nothing
+            return 
+        if self.quiet:
+            self.p = subprocess.Popen(['{}/build/iface/dreami'.format(self.DREAMPATH), self.infile], stderr=subprocess.PIPE, stdout=subprocess.PIPE)
+        else:
+            self.p = subprocess.Popen(['{}/build/iface/dreami'.format(self.DREAMPATH), self.infile], stderr=subprocess.PIPE)
+
+    def hasFinished(self, timeout=1):
+        try:
+            self.stderr_data = self.p.communicate(timeout=timeout)[1].decode('utf-8')
+
+            if self.p.returncode != 0:
+                self.errorOnExit = 1
+            else:
+                self.obj = DREAMOutput(self.outfile)
+
+                if self.deleteOutput:
+                    os.remove(self.outfile)
+        except TimeoutExpired: 
+            # In this case it is expected situation. Process has not completed during specified timeout.
+            # We don't need to preserve any state, because documentation says that no stream data will be lost.
+            return False
+        except KeyboardInterrupt:
+            self.errorOnExit = 2
+
+        # Now we implement timeout manually, by checking start time with current time
+        if time.time() - self.startTime > self.timeout:
+            self.p.kill()
+            self.errorOnExit = 3
```

Comment:
Should this be in the ``except TimeoutExpired`` statement? It seems to me that either (i) the call to ``self.p.communicate()`` would time out and throw a ``TimeoutExpired`` exception, after which this method would now immediately return ``False``, or (ii) the process would complete before the timeout after which ``self.p.kill()`` might be called (if the ``if`` condition is ``True``) on a process which is already finished?

Reply:
The reason the timeout test passes now seems to be that ``self.timeout`` is shorter than the runtime of the simulation, so that ``errorOnExit=3`` after the process is finished. But it is not actually killed before the simulation completes.

Reply:
According to https://docs.python.org/3/library/subprocess.html#subprocess.Popen.communicate TimeoutExpired exception is thrown when timeout of communication method is expired. This is expected, because we want to communicate in round-robin style with all tasks to see if they have completed/crashed. This is NOT a timeout for whole task! The algorithm here is:
1. run N processes.
2. communicate with each of them to see if it completed. Communicate with short timeout. If it throws TimeoutExpired, then assume that the process is still running and just go to checking next process. If process completes/crashes, start next process from input queue.
3. Repeat until input queue is empty.

Reply:
I agree with how ``TimeoutExpired`` works, but the placement of this if statement currently makes ``self.timeout`` have a different effect from what we want it to have. This ``if`` statement will only ever be reached if the process has finished *normally*. If the process took longer than ``self.timeout`` to complete, then ``p.kill()`` will be called after the process completed (on a process which is no longer active) and the task will be indicated as failed. The value of ``self.timeout`` will not however limit for how long the process is allowed to run.

The behaviour we want is instead for ``self.timeout`` to limit for how long a process is allowed to run (sometimes simulations hangs, and so if they run for more than, say 30 minutes, we can consider them failed). For this to happen, this if statement should instead go inside the ``except TimeoutExpired`` block.

(I ran the test and, after inserting som ``print``s, could confirm that this is indeed how the code behaves presently)

Reply:
OH, now I see it. You are right obviously :) I'll fix it in the evening. Thanks :)

Reply:
I fixed it and pushed changes.